### PR TITLE
feat: Improve dev experience by using local S3 

### PR DESCRIPTION
**Pull request by Acruve15**
## Description

Linked to [Issue 217](https://github.com/pyronear/pyro-api/issues/217). Developing locally is not possible if the developer does not have an AWS S3 access. 

## What was done

We used [localstack](https://github.com/localstack) to solve this issue.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,6 +35,7 @@ stop:
 run-dev:
 	docker build src/. -t pyroapi:python3.8-alpine3.10
 	docker-compose -f docker-compose-dev.yml up -d --build
+    awslocal s3api create-bucket --bucket sample-bucket
```
Comment (frgfm) [inline Makefile:38]: slight problem of indentation I think no?

File: client/pyproject.toml [hunk 2]
```diff
@@ -70,6 +70,8 @@ dev = [
     "docutils<0.18",
     "Jinja2<3.1",
     "furo>=2022.3.4",
+    "awscli>=1.27.91",
+    "localstack>=1.4.0"
```
Comment (frgfm) [inline client/pyproject.toml:74]: I think we should also add this to the "test" extra

File: docker-compose-dev.yml [hunk 3]
```diff
@@ -1,6 +1,21 @@
...
+      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
+    environment:
+      - DEBUG=${DEBUG-}
+      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
+      - DOCKER_HOST=unix:///var/run/docker.sock
+      - SERVICES=s3
+      - DATA_DIR=/tmp/localstack/data
+    volumes:
+      - "${LOCALSTACK_VOLUME_DIR:-./volume}:/var/lib/localstack"
+      - "/var/run/docker.sock:/var/run/docker.sock"
```
Comment (frgfm) [inline docker-compose-dev.yml:18]: I cannot really comment on that, perhaps we could add a URL to reference a template docker config for localstack ?

File: docker-compose-dev.yml [hunk 4]
```diff
@@ -1,6 +1,21 @@
 version: '3.7'
 
 services:
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
```
Comment (frgfm) [inline docker-compose-dev.yml:6]: Mind specifying the tag please?

File: docker-compose-dev.yml [hunk 5]
```diff
@@ -18,13 +33,11 @@ services:
...
-      - S3_ACCESS_KEY=${S3_ACCESS_KEY}
-      - S3_SECRET_KEY=${S3_SECRET_KEY}
-      - S3_REGION=${S3_REGION}
+      - BUCKET_NAME=sample-bucket
+      - S3_ENDPOINT_URL=http://s3.localhost.localstack.cloud:4566
+      - S3_ACCESS_KEY=fake
+      - S3_SECRET_KEY=fake
+      - S3_REGION=us-east-1
     depends_on:
       - db
```
Comment (frgfm) [inline docker-compose-dev.yml:42]: Now I think this depends also on localstack

File: docker-compose-dev.yml [hunk 6]
```diff
@@ -1,6 +1,21 @@
...
+    ports:
+      - "127.0.0.1:4566:4566"            # LocalStack Gateway
+      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
+    environment:
+      - DEBUG=${DEBUG-}
+      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
+      - DOCKER_HOST=unix:///var/run/docker.sock
+      - SERVICES=s3
+      - DATA_DIR=/tmp/localstack/data
+    volumes:
```
Comment (frgfm) [inline docker-compose-dev.yml:16]: Shouldn't we add this to the volume declaration at the bottom of the file?

**Review by frgfm — COMMENTED** [review]
Thanks a lot for the PR @Acruve15 :pray: 

I added a few comments but this is looking really good! I have one extra request: is this running in the CI tests?

File: Makefile (hunk 1 above)
Comment (Acruve15) [reply Makefile:38]: weird, I can't seem able to fix it
image

**Comment by Acruve15** [comment]
Still need to solve failing tests.

File: docker-compose-dev.yml (hunk 6 above)
Comment (Acruve15) [reply docker-compose-dev.yml:13]: volumes were updated

**Comment by Acruve15** [comment]
> Thanks a lot for the PR @Acruve15 🙏
> 
> I added a few comments but this is looking really good! I have one extra request: is this running in the CI tests?

Hey @frgfm, localstack was added on part of the tests, tell me if it's good for you, thx

File: .gitignore [hunk 7]
```diff
@@ -147,3 +147,6 @@ client/pyroclient/version.py
 *.db
 _build/
 *.sql
+
+# Docker Volume
+volume/
```
Comment (frgfm) [inline .gitignore:152]: Missing an EOF

**Review by frgfm — APPROVED** [review]
Thank you very much :)

**Review by frgfm — CHANGES_REQUESTED** [review]
My bad there are still a few things to fix the CI:
- we need to integrate that also in the python client test suite
- strangely the test job is failing now :thinking: 

Could you take a look please?

**Comment by frgfm** [comment]
Any update @Acruve15 ? :smiley: 
It would be a shame that we don't use your work!

**Comment by blenzi** [comment]
Hi @Acruve15, very nice addition! I tried to run this locally but `make test` fails on the `docker-compose exec localstack` command. If I issue the command by hand again it works. Any idea ?

**Comment by Acruve15** [comment]
> 

Should be solved

**Review by frgfm — DISMISSED** [review]
The CI is failing because the PR comes from a fork and can't access repo secrets. I've just run everything locally, now everything is in place :smile: 

Thanks a lot @Acruve15 :pray:

**Comment by Acruve15** [comment]
@frgfm thx for merging!

**Pull request merged**